### PR TITLE
rbac: add matching range of destination ports

### LINKED ISSUE #16039: Support destination port range in config.rbac.Permission
I am using [config.rbac.Permission](https://github.com/envoyproxy/envoy/blob/5a8bfa20dc3c85ecb61826d122696ecaa75dffa0/api/envoy/config/rbac/v3/rbac.proto#L145) to configure RBAC for a dynamic forward proxy.

My proxy configuration needs to support allowing large port ranges.

Today, I can achieved this by combining thousands of [Permission.destination_port](https://github.com/envoyproxy/envoy/blob/5a8bfa20dc3c85ecb61826d122696ecaa75dffa0/api/envoy/config/rbac/v3/rbac.proto#L182) rules using [Permission.Set.or_rules](https://github.com/envoyproxy/envoy/blob/5a8bfa20dc3c85ecb61826d122696ecaa75dffa0/api/envoy/config/rbac/v3/rbac.proto#L164),
similar to this (modified) snippet from the [config.rbac.RBAC docs](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/rbac/v3/rbac.proto#config-rbac-v3-rbac).

```
  "product-viewer":
    permissions:
        - and_rules:
            rules:
              - header: { name: ":method", exact_match: "GET" }
              - url_path:
                  path: { prefix: "/products" }
              - or_rules:
                  rules:
                    - destination_port: 1000
                    - destination_port: 1001
                    # .
                    # ... a lot of lines
                    # .
                    - destination_port: 1999
```

This is functional but cumbersome.

It would be more ergonomic if Permission had a destination port range field, similar to what exists in [config.listener.ListenerFilterChainMatchPredicate](https://github.com/envoyproxy/envoy/blob/5a8bfa20dc3c85ecb61826d122696ecaa75dffa0/api/envoy/config/listener/v3/listener_components.proto#L324).

### PR DESCRIPTION
**Pull request by zhxie**
Signed-off-by: Xie Zhihao <zhihao.xie@intel.com>

Commit Message: rbac: add matching range of destination ports
Additional Description: use one destination port range rule to cover thousands of port rules in RBAC permissions, adding convenience for writing configuration and making matching quicker
Risk Level: Low
Testing: Unit
Docs Changes: N/A
Release Notes: Added
Platform Specific Features: N/A
Fixes #16039

### CODE REVIEW

**Comment by leyao-daily** [comment]
please rebase and sign all your commits with `git commit -s` to pass DCO check.

File: api/envoy/config/rbac/v3/rbac.proto
Comment (adisuissa) [inline api/envoy/config/rbac/v3/rbac.proto:187]: A few questions:
1. Are there any constraints on the values here? Is there a default value if it is not specified?
2. How does this interact with `destination_port` above? Should the previous field be deprecated?
3. Could there be a use-case of multiple ranges (should this be a repeated field)?

**Review by adisuissa — COMMENTED** [review]
Thanks for working on this!
Left a few API questions.

File: api/envoy/config/rbac/v3/rbac.proto
Comment (zhxie) [reply api/envoy/config/rbac/v3/rbac.proto:187]: 1. The `destination_port_range` in RBAC is the same with `destination_port_range` in listener filter chain match predicate. Maybe it will be natural to add a constraint where the port must between 0 and 65535, and the `start` port must be less than the `end` port, but an input like [443, 443), [443, 80) or [-1, 65537) will not actually cause any exceptions, the matcher still works as usual. And Also, there is no default value.

https://github.com/envoyproxy/envoy/blob/9d0f0355135a19dff6949fed884e81b0d1eb101a/api/envoy/config/listener/v3/listener_components.proto#L330-L332

2. The `destination_port` and `destination_port_range` can work together using logic rule `or_rules` in RBAC. For example, a rule set like

   ```yaml
   - or_rules:
       rules:
         - destination_port_range:
             start: 21
             end: 23
         - destination_port: 80
   ```

   will filter out connections except there destination port is 21, 22 or 80. I am not sure whether the `destination_port` should be deprecated or not, since I am new to Envoy, and not aware of users' request.

3. Yes. A rule set like

   ```yaml
   - or_rules:
       rules:
         - destination_port_range:
             start: 21
             end: 23
         - destination_port_range:
             start: 30000
             end: 40000
   ```

   is valid.

Comment (adisuissa) [reply api/envoy/config/rbac/v3/rbac.proto:187]: @htuch for another set of eyes on this:
1. Should `destination_port_range` deprecate the `destination_port`?
2. Do you think making the new field repeated makes sense?

Comment (htuch) [reply api/envoy/config/rbac/v3/rbac.proto:187]: 1. If we were designing from scratch, I might be tempted to go there, but given the material UX simplification for a common case and the churn of deprecation, I think it's fine to say what is here is fine. Range matches are likely to be rarer than singleton (or we would have seen this PR by now).
2. Don't feel super strongly either way, if it's not going to be very common to be punching lots of holes in ranges then a single int range with RBAC conjunction/disjunction rules is going to be fine.

Comment (zhxie) [reply api/envoy/config/rbac/v3/rbac.proto:187]: 1. In fact, I also agree with @htuch that the `destination_port_range` should not be deprecate `destination_port` in consideration of compatibility. The purpose of `destination_port_range` is introduced by #16039 which using large amount of ports for dynamic forward proxy, and not all the scene want to replace the single-line `destination_port` with the 3-line `destination_port_range`.
2. `destination_port_range` itself is not a repeated field, but all the rule in RBAC permission can be repeated with the `or_rules`.

Comment (adisuissa) [reply api/envoy/config/rbac/v3/rbac.proto:190]: I see, thanks for explaining this.

**Review by adisuissa — COMMENTED** [review]
/lgtm api

**Comment by htuch** [comment]
@adisuissa could you also take a look at the first pass for implementation review when you're back? Thanks!

File: test/extensions/filters/common/rbac/matchers_test.cc
Comment (adisuissa) [inline test/extensions/filters/common/rbac/matchers_test.cc:110]: nit: why not copy `range` into `destination_port_range`?

File: test/extensions/filters/common/rbac/matchers_test.cc
Comment (adisuissa) [inline test/extensions/filters/common/rbac/matchers_test.cc:246]: Please add a short comment stating what the test validates.

**Review by adisuissa — COMMENTED** [review]
Implementation and tests look good overall, thanks!
Left a couple of nits.

File: test/extensions/filters/common/rbac/matchers_test.cc
Comment (zhxie) [reply test/extensions/filters/common/rbac/matchers_test.cc:107]: That's right. Unused variable has been removed.

File: test/extensions/filters/common/rbac/matchers_test.cc
Comment (zhxie) [reply test/extensions/filters/common/rbac/matchers_test.cc:244]: Ok, comment added.

**Review by adisuissa — APPROVED** [review]
Thanks, LGTM.
@htuch for the next pass

File: source/extensions/filters/common/rbac/matchers.h
Comment (htuch) [inline source/extensions/filters/common/rbac/matchers.h:169]: Arguably we should reject negative and OOB values here for ports.

**Review by htuch — CHANGES_REQUESTED** [review]
LGTM modulo small ask.
/wait

File: source/extensions/filters/common/rbac/matchers.h
Comment (zhxie) [reply source/extensions/filters/common/rbac/matchers.h:169]: To validate the range of ports, we are better to add a validation rule in RBAC proto. 

https://github.com/envoyproxy/envoy/blob/f97bdf143f570e10bd919369235fc801f3a65fd2/api/envoy/config/rbac/v3/rbac.proto#L189-L190

But the type `type.v3.Int32Range` does not have any validation rules, and only certain types have validation rules that is located in [validate.proto](https://github.com/envoyproxy/protoc-gen-validate/blob/main/validate/validate.proto#L36-L66), which belongs to another repo [envoyproxy/protoc-gen-validate](https://github.com/envoyproxy/protoc-gen-validate). As a result, we are not easy to reject invalid or illegal values.

In addition, `type.v3.Int32Range` is also used in destination port matching in listener filter, but the `ListenerFilterDstPortMatcher` does not check its input.

https://github.com/envoyproxy/envoy/blob/f97bdf143f570e10bd919369235fc801f3a65fd2/api/envoy/config/listener/v3/listener_components.proto#L330-L332

https://github.com/envoyproxy/envoy/blob/f97bdf143f570e10bd919369235fc801f3a65fd2/source/common/network/filter_matcher.h#L40-L46

So I think maybe we do not have to add a check?

Comment (htuch) [reply source/extensions/filters/common/rbac/matchers.h:169]: Yes, this is a limit of PGV annotations and proto typing, I don't think it can be done cleanly at the proto annotation level. I think both `ListenerFilterDstPortMatcher` and your code could share a common utility that at config ingestion time, checks the port range and throws if invalid; this is the next best thing.

**Review by htuch — APPROVED** [review]
LGTM, thanks!

**Pull request merged**
